x/wasm/internal/keeper/keeper.go

```diff
@@ -693,6 +694,54 @@ func (k Keeper) dispatchMessages(ctx sdk.Context, contractAddr sdk.AccAddress, i
 	return nil
 }
 
+func (k Keeper) PinContract(ctx sdk.Context, contractAddr sdk.AccAddress) error {
+	_, codeInfo, _, err := k.contractInstance(ctx, contractAddr)
+	if err != nil {
+		return err
+	}
+	if err := k.wasmer.Pin(codeInfo.CodeHash); err != nil {
+		return sdkerrors.Wrap(types.ErrPinContractFailed, err.Error())
+	}
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.GetPinnedContractIndexPrefix(contractAddr), []byte{})
+	return nil
+}
+
+func (k Keeper) UnpinContract(ctx sdk.Context, contractAddr sdk.AccAddress) error {
+	_, codeInfo, _, err := k.contractInstance(ctx, contractAddr)
+	if err != nil {
+		return err
+	}
+	if err := k.wasmer.Unpin(codeInfo.CodeHash); err != nil {
+		return sdkerrors.Wrap(types.ErrUnpinContractFailed, err.Error())
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.GetPinnedContractIndexPrefix(contractAddr))
+	return nil
+}
+
+func (k Keeper) IsPinnedContract(ctx sdk.Context, contractAddr sdk.AccAddress) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Has(types.GetPinnedContractIndexPrefix(contractAddr))
+}
+
+// InitializePinnedContracts updates wasmvm to pin to cache all contracts marked as pinned
+func (k Keeper) InitializePinnedContracts(ctx sdk.Context) error {
```

Comment:
Ah, this is the difference, you don't need to call `Pin` everytime, but you do read the store to check for gas.

Yeah, this works well. And we can try to cache the state for `IsPinnedContract` query in 0.17